### ISSUE TITLE
Verbose Option

### ISSUE THREAD

**Issue by OlivierLabo** [JKISoftware/Caraya#43]
It could be nice to add a "Verbose" option on assertions VIs.
Particularly when using tests reports and being integrated in a CI environment.

For instance, when an Equal (Value) assertion fails, currently we only have the title we put to the assertion function 
14:14:56 02/19/2019 - PASS -       Data Read = Expected data
14:14:56 02/19/2019 - FAIL -       Cmd Type = Expected
![image](https://user-images.githubusercontent.com/41787807/53022414-6b659b00-345b-11e9-9cc6-fec976284306.png)
There could be an additional information displaying expected value and measured value
14:14:56 02/19/2019 - PASS -       Data Read = Expected data
14:14:56 02/19/2019 - FAIL -       Cmd Type = Expected      - {Expected : 2 ; Measured : 4}

Then, to adapt the same code to all function, maybe the datatype would be needed.
Of course, there will be a limitation to usual datatypes ( I think it's not really possible for objects).

**Comment by jimkring** [comment]
@OlivierLabo I like the idea of automatically generating a descriptive failure reason/string that gets displayed on the UI and in the report somehow. Certainly, this makes it much easier to debug why the test failed. It's also nice to have a continuous integration server email you this information, rather than opening the code up and trying to run the test, in order to probe the wires :-)

**Comment by francois-normandin** [comment]
@OlivierLabo 
I believe there is a way to add this functionality, but because it can have an impact on execution speed, I would make it verbose only if the user specifically configures a Test Suite with a "verbose" option flag set to True. We can't simply add another flag as a terminal on all the Assertions: it is getting cluttered and not all assertions still have room available on their connector pane. 

I've added this feature request to the roadmap for release with Caraya 1.0 so work can start on it, but this is contingent on initial tests.

The roadmap for implementing the feature is to add a flag to the Test Suite, much in the same way that one can inject a custom Test Report. Once the suite is configured, the verbose mode can be implemented under the hood without the user worrying about further verbose flags. I'll have a proposed solution in the coming days.

**Comment by francois-normandin** [comment]
Alright, I have a proof of concept with the Assert True case. The verbose mode works for Test Suites and will not impact the execution speed if verbose mode is Off.

![image](https://user-images.githubusercontent.com/11728548/70495684-50dcdb00-1adc-11ea-929e-f0e160115f05.png)

This is the "Assert True" node:

![image](https://user-images.githubusercontent.com/11728548/70495588-fc396000-1adb-11ea-8766-d6d0fa3b2e00.png)

Report looks like this (Simple Text format):

![image](https://user-images.githubusercontent.com/11728548/70495565-ef1c7100-1adb-11ea-98a9-8311cc024dfd.png)

To be done: 
- transpose to all other assertion nodes based on the inputs (variant to string, essentially)
- add an optional property field in the JUnit report when verbose = true

**Comment by francois-normandin** [comment]
Done.
All assertions now have a proper failure message. 
- Cluster equality assertion will return the first element that is not equal.
- Error assertion will return a warning code if code !=0

This addition required that I use OpenG String package as a new dependency. It could be avoided for simple datatypes, but there was no point in reinventing the wheel for more complex types such as clusters or arrays.

> 2019-12-19 09:17:50 - FAIL - Test Report Generation.vi - Test Report Generation.vi
> 1903-12-31 19:00:00 - FAIL -   Verbosity Test.vi - Verbosity Test.vi
> 2019-12-19 09:17:50 - FAIL -     Verbose Equal Cluster - {Cluster element index: 1 -- Expected value: 41, Asserted value: 42}
> 2019-12-19 09:17:50 - FAIL -     Verbose False - {Expected: False, Asserted True}
> 2019-12-19 09:17:50 - FAIL -     Verbose True - {Expected: True, Asserted: False}
> 2019-12-19 09:17:50 - FAIL -     Verbose Error - {Expected: Error, Asserted: Warning 42}
> 2019-12-19 09:17:50 - FAIL -     Verbose No Error - {Expected: No Error, Asserted: error code = 42}
> 2019-12-19 09:17:50 - FAIL -     Verbose Greater Value - {Expected: "3"  to be greater than "4.000000", Asserted: less or equal.}
> 2019-12-19 09:17:50 - FAIL -     Verbose Equal Value - {Expected value: 3, Asserted value: 4.000000}
> 2019-12-19 09:17:50 - FAIL -     Verbose Less or Equal Value - {Expected: "5" to be smaller or equal to "4.000000", Asserted: greater}

**Comment by OlivierLabo** [comment]
Nice job François.
Actually I found the text quite long, but it has the merit of being really clear without ambiguity (if the label of assertion were not that clear for instance)
I really like the way you manage Clusters and errors.

**Comment by francois-normandin** [comment]
@OlivierLabo  I'd like to hear your thoughts/proposal for a more concise description. The shorter the message is, the less annoying it will be when used repetitively.
Of course, only parallel assertions will show up as failures with a description because downstream assertions will simply skip.

**Comment by OlivierLabo** [comment]
François, be sure that if I had a valuable proposal, I would have told you.
I was thinking of removing the expected result for obvious assertions like True or False. But Actually, what appears as "Verbose False" in the Log is only the label you decided to assign to this assertion. If some use a more "functional" label, we loose the knowledge of the assertion implemented. Your solution permits to be really clear in all circumstances.
Or maybe we have to find the used assertion directly in the "expected" field.
For instance, for True assertion, remove the asserted value that is obvious as the verbose display is only for failed test. If expected is True, we know that asserted value is False.
For greater assertions, I was thinking of something like {Expected: Greater than 4; Asserted : 3} but that means that there is an input specified as "Expected" and the other as "Asserted". I'm not sure it suits well all usages.

**Comment by francois-normandin** [comment]
> For greater assertions, I was thinking of something like {Expected: Greater than 4; Asserted : 3} but that means that there is an input specified as "Expected" and the other as "Asserted". I'm not sure it suits well all usages.

Indeed, that was my initial thought too. I had tried to make it as you describe and then realized that the two elements being compared could be reversed and who knows what is the "expected" value.

I still have another change I want to implement before going into formal review for branch 1.0, so I'll see if others want to chip in with suggestions. In truth, I would have liked to make this dependency-injectable (with a default proposition) but the current framework does not allow it as easily as it was for reports.
